lib/vector-core/src/transform/mod.rs

```diff
@@ -382,54 +407,81 @@ impl ByteSizeOf for TransformOutputsBuf {
 }
 
 #[derive(Debug, Default)]
-pub struct OutputBuffer(Vec<Event>);
+pub struct OutputBuffer(Vec<EventArray>);
 
 impl OutputBuffer {
     pub fn with_capacity(capacity: usize) -> Self {
         Self(Vec::with_capacity(capacity))
     }
 
     pub fn push(&mut self, event: Event) {
-        self.0.push(event);
+        // Coalesce multiple pushes of the same type into one array.
+        match (event, self.0.last_mut()) {
+            (Event::Log(log), Some(EventArray::Logs(logs))) => {
+                logs.push(log);
+            }
+            (Event::Metric(metric), Some(EventArray::Metrics(metrics))) => {
+                metrics.push(metric);
+            }
+            (event, _) => {
+                self.0.push(event.into());
+            }
+        }
     }
 
     pub fn append(&mut self, events: &mut Vec<Event>) {
-        self.0.append(events);
+        for event in events.drain(..) {
```

Comment:
Curious about this `drain`, why can't we use `for event in events` as below?

Reply:
Because `append` doesn't own the events, it only has a reference.

Reply:
Would it make sense to consume and own the `events` here instead? Or what are the benefits of leaving the provided `events` empty?

Reply:
Maybe. I was going on leaving the interface for the `OutputBuffer` the same as before as much as possible to reduce the change surface. Note that `fn Vec::append` has a similar interface to this (it takes an `&mut Vec<T>` and drains it). I wouldn't expect it to have much of a performance impact in any case.

Reply:
Interesting, I wouldn't have designed it that way in the first place, but following the conventions of the stdlib is a strong argument. Wouldn't expect this to have any performance implication either, was just worried about the ergonomics of this.